cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,280 @@
+
+package release
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	kcdatav1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
+	yaml2 "gopkg.in/yaml.v2"
+	yaml3 "gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+// keys used when generating an OpenAPI Document
+const (
+	typeKey        = "type"
+	formatKey      = "format"
+	descriptionKey = "description"
+	itemsKey       = "items"
+	propertiesKey  = "properties"
+	defaultKey     = "default"
+)
+
+var keyOrder = map[string]int{
+	typeKey:        1,
+	formatKey:      2,
+	descriptionKey: 3,
+	itemsKey:       4,
+	propertiesKey:  5,
+	defaultKey:     6,
+}
+
+const (
+	objectVal = "object"
+	arrayVal  = "array"
+	floatVal  = "float"
+)
+
+// YAML tags that differentiate the type of scalar object in the node
+const (
+	nullTag  = "!!null"
+	boolTag  = "!!bool"
+	intTag   = "!!int"
+	floatTag = "!!float"
+)
+
+type MapItem struct {
+	Key   string
+	Value interface{}
+}
+
+type Map struct {
+	Items []*MapItem
+}
+
+func (m *Map) Iterate(iterFunc func(k string, v interface{})) {
+	for _, item := range m.Items {
+		iterFunc(item.Key, item.Value)
+	}
+}
+
+type HelmValuesSchemaGen struct {
+	// dirPath to the helm chart directory
+	dirPath string
+}
+
+func NewHelmValuesSchemaGen(path string) HelmValuesSchemaGen {
+	return HelmValuesSchemaGen{path}
+}
+
+func (h HelmValuesSchemaGen) Schema() (*kcdatav1alpha1.ValuesSchema, error) {
+	fileData, err := h.readValuesFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(fileData) == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	var document yaml3.Node
+	err = yaml3.Unmarshal(fileData, &document)
+	if err != nil {
+		return nil, err
+	}
+	if document.IsZero() {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	if document.Kind != yaml3.DocumentNode {
+		return nil, fmt.Errorf("invalid node kind supplied: %d", document.Kind)
+	}
+	if document.Content[0].Kind != yaml3.MappingNode {
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.toYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	fileInfo, err := os.Stat(h.dirPath)
+	if err != nil {
+		return nil, err
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("expected %s to be directory", h.dirPath)
+	}
+
+	fileData, err := os.ReadFile(filepath.Join(h.dirPath, "values.yaml"))
```

Comment:
Should we also account for `values.yml` or is it always `values.yaml`?

Reply:
We should probably see whether helm accepts both

Reply:
Helm doesnt accept `values.yml`.